### PR TITLE
API tests for shipping zones and methods

### LINKED ISSUE #30676: Write new API test: Shipping zones > Can add shipping zones

### PR DESCRIPTION
**Pull request by rodelgc**
### All Submissions:

* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?

### Changes proposed in this Pull Request:

Closes #30676 and #30677.

### How to test the changes in this Pull Request:

1. Clone this branch and run the setup scripts for the e2e environment.
2. Run the new tests for shipping zones and shipping methods:
```
npm test -- --group=shipping-zones
npm test -- --group=shipping-methods
```
3. You should see the `Shipping zones API tests` and `Shipping methods API tests` pass:
```
 PASS  tests/shipping/shipping-zones.test.js
  Shipping zones API tests
    ✓ can create a shipping zone (243ms)
    ✓ can retrieve a shipping zone (88ms)
    ✓ can list all shipping zones (93ms)
    ✓ can update a shipping zone (89ms)
    ✓ can delete a shipping zone (175ms)

 PASS  tests/shipping/shipping-method.test.js
  Shipping methods API tests
    ✓ can add a 'Flat rate' shipping method (217ms)
    ✓ can add a 'Free shipping' shipping method (181ms)
    ✓ can add a 'Local pickup' shipping method (192ms)
```

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Added: Shipping Zones API Tests
> Added: Shipping Methods API Tests

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by zhongruige** [comment]
@rodelgc Looks like we're seeing a similar failure to when the API tests are running after the full e2e test suite--I am curious if you are seeing the same thing locally when running these after the e2e test suite, e.g.:

```
pnpx wc-e2e test:e2e
pnpx wc-api-tests test api
```

**Comment by rodelgc** [comment]
Oh right, I'm also getting the same assertion error after the full e2e suite ran. The e2e suite seems to leave a few shipping zones undeleted. I'll update the assertion for now.

**Comment by rodelgc** [comment]
Done. I removed the flaky assertion in this commit: https://github.com/woocommerce/woocommerce/pull/31141/commits/756644dac33e332acee9e6d6f3736f0d3b86befd

**Comment by zhongruige** [comment]
> Done. I removed the flaky assertion in this commit: [756644d](https://github.com/woocommerce/woocommerce/commit/756644dac33e332acee9e6d6f3736f0d3b86befd)

Thanks for getting this in @rodelgc!

Thinking about this more just now, I think rather than adjust the assertions (we should be verifying the order response length, I think, as that is a good test-case), we should adjust the API tests to run in a separate job after the e2e tests have ran.

As a result, I decided to submit a PR to run the tests separate from the e2e tests to avoid this flakiness in the future: https://github.com/woocommerce/woocommerce/pull/31218

Screen Shot 2021-11-16 at 09 14 01

Unfortunately it looks like the tests on that PR are failing due to other unrelated issues:

```
/home/runner/work/woocommerce/woocommerce/package/woocommerce/packages/js/api/dist/index.js:1
export { HTTPClientFactory } from './http';
^^^^^^

SyntaxError: Unexpected token 'export'
```

If you run `pnpm install` and try and bring Docker up on `trunk` you'll see the same.

Anyway, I am thinking we put the length check back in that was removed on [756644d](https://github.com/woocommerce/woocommerce/commit/756644dac33e332acee9e6d6f3736f0d3b86befd) and we either wait for the other PR to be merged in or go ahead with these changes as they are.

**Comment by rodelgc** [comment]
Thanks @zhongruige!

What if I'll add the length assertion back, but in addition to it I'll add a step in the `beforeAll` block to batch-delete all pre-existing shipping zones?

This way, we can safely merge this PR without depending on [Run API tests in their own job](https://github.com/woocommerce/woocommerce/pull/31218)

`tests/shipping/shipping-zones.test.js` is the only Core API test (so far) that uses shipping zones. And that `tests/shipping/shipping-method.test.js` uses the default, unremovable zone "Locations not covered by your other zones" anyway. So maybe batch-deleting pre-existing zones at the beginning of the test would be fine?

**Comment by zhongruige** [comment]
> What if I'll add the length assertion back, but in addition to it I'll add a step in the beforeAll block to batch-delete all pre-existing shipping zones?

I think that should work and is a good way to start the test clean.

Just as a side note this PR is now in review and the pipeline passed with the changes: https://github.com/woocommerce/woocommerce/pull/31218

**Comment by rrennick** [comment]
```
/home/runner/work/woocommerce/woocommerce/package/woocommerce/packages/js/api/dist/index.js:1
export { HTTPClientFactory } from './http';
^^^^^^

SyntaxError: Unexpected token 'export'
```

@rodelgc @zhongruige That was [this issue](https://github.com/woocommerce/woocommerce/issues/31223).

File: packages/js/api-core-tests/tests/shipping/shipping-zones.test.js
Comment (zhongruige) [inline packages/js/api-core-tests/tests/shipping/shipping-zones.test.js]: It might be slightly an edge case, but it may be worth adding some tests around the following:

* can't delete the default shipping zone
* the default shipping zone exists and is the only option before we add more

That way we can get coverage if these ever changes, especially since we have some checks in the e2e-utils, for example, expecting this behavior:

https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/e2e-utils/src/flows/with-rest-api.js#L275

**Review by zhongruige — COMMENTED** [review]
Left a recommendation for some tests we can add around the default shipping zone.

File: packages/js/api-core-tests/tests/shipping/shipping-zones.test.js
Comment (rodelgc) [reply packages/js/api-core-tests/tests/shipping/shipping-zones.test.js]: Thanks @zhongruige! Yep, makes sense. 

I have just added a test to make sure that the default shipping zone remains after deleting all zones. 
I also added a test to make sure that we don't allow it to be modified by the user.

Comment (zhongruige) [reply packages/js/api-core-tests/tests/shipping/shipping-zones.test.js]: Thanks for adding that in!

File: packages/js/api-core-tests/endpoints/shipping-methods.js
Comment (zhongruige) [inline packages/js/api-core-tests/endpoints/shipping-methods.js]: This should be "Delete a shipping method from a shipping zone".

**Review by zhongruige — CHANGES_REQUESTED** [review]
These ran great for me locally:

```
 PASS  tests/shipping/shipping-zones.test.js
  Shipping zones API tests
    ✓ cannot delete the default shipping zone "Locations not covered by your other zones" (278ms)
    ✓ cannot update the default shipping zone (87ms)
    ✓ can create a shipping zone (91ms)
    ✓ can retrieve a shipping zone (90ms)
    ✓ can list all shipping zones (100ms)
    ✓ can update a shipping zone (95ms)
    ✓ can delete a shipping zone (188ms)

 PASS  tests/shipping/shipping-method.test.js
  Shipping methods API tests
    ✓ can add a 'Flat rate' shipping method (208ms)
    ✓ can add a 'Free shipping' shipping method (187ms)
    ✓ can add a 'Local pickup' shipping method (197ms)
```

Just one minor change for the delete shipping method from shipping zone. Once that's resolved this should be good to merge.

File: packages/js/api-core-tests/endpoints/shipping-methods.js
Comment (rodelgc) [reply packages/js/api-core-tests/endpoints/shipping-methods.js]: Woops, sorry about that! I had just pushed the correction now. Thanks again!

Comment (zhongruige) [reply packages/js/api-core-tests/endpoints/shipping-methods.js]: No problem! Looks good!

**Review by zhongruige — DISMISSED** [review]
Looks good to me! Once the pipeline gets sorted this is good to merge.

**Comment by zhongruige** [comment]
Hi @rodelgc!

Just wanted to let you know the [API core tests package was recently released](https://www.npmjs.com/package/@woocommerce/api-core-tests) 🎉  So we'll just need to move the changelog entries up to an unreleased section so it can be included in a future release:

```
- Shipping Zones API Tests
- Shipping Methods API Tests
```

**Comment by rodelgc** [comment]
Thanks @zhongruige!

> we'll just need to move the changelog entries up to an unreleased section 

Got it! I'll do the same with the complex order PR.
Seems like `trunk` is in a much better state now. I'll merge trunk to this branch as well.

**Comment by zhongruige** [comment]
@rodelgc Looks like the e2e tests failed, just going to re-run them to see if they pass.

**Comment by zhongruige** [comment]
Changes look good! Thanks for making the changelog updates!

:shipit:

**Pull request merged**
